apps/test/ChartApiTest.js

```diff
@@ -0,0 +1,367 @@
+    assert.isTrue(ChartApi.supportsType(ChartApi.ChartType.SCATTER));
+    assert.isTrue(ChartApi.supportsType('SCATTER'));
+    assert.isTrue(ChartApi.supportsType('Scatter'));
+    assert.isTrue(ChartApi.supportsType('scatter'));
+  });
+
+  it("quotes and alphabetizes types for dropdown", function () {
+    assert.deepEqual(ChartApi.getChartTypeDropdown(), [
+        '"bar"',
+        '"line"',
+        '"pie"',
+        '"scatter"'
+    ]);
+  });
+
+  describe("drawChartFromRecords", function () {
+    var chartApi, result, rejection;
+
+    beforeEach(function () {
+      chartApi = new ChartApi(fakeDocument, fakeAppStorage);
+      result = null;
+      rejection = null;
+    });
+
+    function onResolve(r) {
+      result = r;
+    }
+
+    function onReject(e) {
+      rejection = e;
+    }
+
+    var assertWarns = function (chartApi, warningRegexp) {
+      assert.isNull(rejection);
+      var warningFound = chartApi.warnings.some(function (e) {
+        return warningRegexp.test(e.message);
+      });
+      assert(warningFound, 'Expected warning ' +
+          warningRegexp.toString() + "\n\Got warnings:\n" +
+          chartApi.warnings.map(function (e) {
+            return e.message;
+          }).join("\n"));
+    };
+
+    var assertNotWarns = function (chartApi, warningRegexp) {
+      assert.isNull(rejection);
+      var warningFound = chartApi.warnings.some(function (e) {
+        return warningRegexp.test(e.message);
+      });
+      assert(!warningFound, 'Expected no warning ' +
+          warningRegexp.toString() + "\n\Got warnings:\n" +
+          chartApi.warnings.map(function (e) {
+            return e.message;
+          }).join("\n"));
+    };
+
+    it("returns a Promise", function () {
+      assert.instanceOf(chartApi.drawChartFromRecords(), Promise);
+    });
+
+    it ("rejects if element is not found", function (testDone) {
+      chartApi.drawChartFromRecords('missingId')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unable to render chart into element "missingId".');
+          }));
+    });
+
+    it ("rejects if element is wrong type", function (testDone) {
+      chartApi.drawChartFromRecords('fakeImg')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unable to render chart into element "fakeImg".');
+          }));
+    });
+
+    it ("rejects if chart type is not supported", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', 'badType')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unsupported chart type "badType".');
+          }));
+    });
+
+    it ("rejects if no columns array provided", function (testDone) {
```

Comment:
I would propose changing this line to "rejects if no columns array provided and no columns found in the data". Otherwise, this test is confusing because it suggests that `columns` must always be present, whereas in reality `columns` do not need to be present if they can be inferred from the data.